go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -364,3 +631,562 @@ func (vc *vcopier) fastForward(ctx context.Context, copyState map[string]*sqltyp
+	return vct.args
+}
+
+func (vct *vcopierCopyTask) ID() int {
+	return vct.id
+}
+
+func (vct *vcopierCopyTask) Lifecycle() *vcopierCopyTaskLifecycle {
+	return vct.lifecycle
+}
+
+// Lastpk returns the lastpk to use to update _vt.copy_state.
+func (vct *vcopierCopyTaskArgs) Lastpk() *querypb.Row {
+	return vct.lastpk
+}
+
+// Rows returns the rows to be inserted into the target table.
+func (vct *vcopierCopyTaskArgs) Rows() []*querypb.Row {
+	return vct.rows
+}
+
+// After returns a vcopierCopyTaskHooks that can be used to register callbacks
+// to be triggered after the task successfully advances past to the specified
+// vcopierCopyTaskState.
+func (vtl *vcopierCopyTaskLifecycle) After(state vcopierCopyTaskState) *vcopierCopyTaskHooks {
+	key := "after:" + state.String()
+	if _, ok := vtl.hooks[key]; !ok {
+		vtl.hooks[key] = newVCopierCopyTaskHooks()
+	}
+	return vtl.hooks[key]
+}
+
+// Before returns a vcopierCopyTaskHooks that can be used to register callbacks
+// to be triggered before the task advances to the specified
+// vcopierCopyTaskState.
+func (vtl *vcopierCopyTaskLifecycle) Before(state vcopierCopyTaskState) *vcopierCopyTaskHooks {
+	key := "before:" + state.String()
+	if _, ok := vtl.hooks[key]; !ok {
+		vtl.hooks[key] = newVCopierCopyTaskHooks()
+	}
+	return vtl.hooks[key]
+}
+
+// OnResult returns a vcopierCopyTaskResultHooks that can be used to register
+// callbacks to be triggered when a task advances to a "done" state (=
+// canceled, completed, failed).
+func (vtl *vcopierCopyTaskLifecycle) OnResult() *vcopierCopyTaskResultHooks {
+	return vtl.resultHooks
+}
+
+// TryAdvance is a convenient way of wrapping up lifecycle hooks with task
+// execution steps. E.g.:
+//
+//	if _, err := task.Lifecycle().Before(nextState).Notify(ctx, args); err != nil {
+//		return err
+//	}
+//	if _, err := fn(ctx, args); err != nil {
+//		return err
+//	}
+//	if _, err := task.Lifecycle().After(nextState).Notify(ctx, args); err != nil {
+//		return err
+//	}
+//
+// Is roughly equivalent to:
+//
+//	if _, err := task.Lifecycle.TryAdvance(ctx, args, nextState, fn); err != nil {
+//	  return err
+//	}
+func (vtl *vcopierCopyTaskLifecycle) TryAdvance(
+	ctx context.Context,
+	args *vcopierCopyTaskArgs,
+	nextState vcopierCopyTaskState,
+	fn func(context.Context, *vcopierCopyTaskArgs) error,
+) (newState vcopierCopyTaskState, err error) {
+	// Canceling the task when there is a context error seems like a sensible
+	// thing to do. E.g. when copyPhaseDuration elapses it will cause the task
+	// to be canceled. Hower this is more aggressive cancelation behavior than
+	// the unit tests currently support. Will need to rework unit tests before
+	// we enable this.
+
+	//defer func() {
+	//	if err != nil {
+	//		//if errors.Is(err, context.Canceled) ||
+	//		//	errors.Is(err, context.DeadlineExceeded) {
+	//		//	newState = vcopierCopyTaskCancel
+	//		//} else {
+	//		//	newState = vcopierCopyTaskFail
+	//		//}
+	//	}
+	//}()
```

Comment:
Is not enabling this going to cause potential PRS/ERS failures? We need to be able to cancel and close things fairly quickly in order to prevent PRS failures.

Reply:
My thinking is that this doesn't change the way things currently work in `main`, which is that if a copy task is inflight, it has to either complete or fail. A context expiration doesn't actually cancel an inflight task.

Reply:
But the better answer is we totally can and should enable this. We will just need to re-worker the tests in `vcopier_test.go`. Will talk through that when we sync.

Reply:
I've added this back. Turns out it doesn't break tests, I must have misunderstood the test failures I was getting when I last had this code active.